### PR TITLE
Loading mistral reward model checkpoints

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
See https://github.com/pytorch/torchtune/issues/812#issuecomment-2067790583 for context.

#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?

I've added some lightweight logic in `torchtune/utils/_checkpointing/_checkpointer.py` to support loading a mistral reward model [from huggingface](https://github.com/huggingface/transformers/blob/1e05671d213642a67dd418c13a18a17a18c25117/src/transformers/models/mistral/modeling_mistral.py#L1282), and added a test case to ensure model weights are converted correctly.

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add unit tests for any new functionality
- [ ] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

### CODE REVIEW

File: torchtune/utils/_checkpointing/_checkpointer_utils.py [hunk 1]
```diff
@@ -24,6 +24,7 @@ class ModelType(Enum):
     LLAMA3 = "llama3"
     MISTRAL = "mistral"
     PHI3_MINI = "phi3_mini"
+    MISTRAL_REWARD = "mistral_reward"
```
Comment (SalmanMohammadi) [inline torchtune/utils/_checkpointing/_checkpointer_utils.py:27]: maybe this could be mistral_classifier?

File: torchtune/utils/_checkpointing/_checkpointer.py [hunk 2]
```diff
@@ -384,6 +384,14 @@ def load_checkpoint(self) -> Dict[str, Any]:
 
         if self._model_type == ModelType.PHI3_MINI:
             converted_state_dict[utils.MODEL_KEY] = phi3_hf_to_tune(merged_state_dict)
+        elif self._model_type == ModelType.MISTRAL_REWARD:
+            merged_state_dict["lm_head.weight"] = merged_state_dict.pop("score.weight")
```
Comment (kartikayk) [inline torchtune/utils/_checkpointing/_checkpointer.py:388]: It does seem very tempting to add this one line change here. But this will start to add up and make this checkpointer really hard to read.

I do need to redesign this component a bit. But in the mean time sn alternative here (granted it's more lines of code with some copy paste) would be to define a custom ```convert_weights.py``` within the mistral folder and define a new ```mistral_reward_hf_to_tune``` and ```mistral_reward_tune_to_hf``` functions which do this mapping (and any future changes you might need for reward modeling). An example is what we did for [```phi3```](https://github.com/pytorch/torchtune/blob/main/torchtune/models/convert_weights.py). As I mentioned, its more copy-paste code, but I can definitely see this function updating as you add more heads etc. And I'm afraid the checkpointer will bloat up because of it.

@ebsmothers let me know what you think about this.

File: torchtune/utils/_checkpointing/_checkpointer.py [hunk 3]
```diff
@@ -426,6 +434,16 @@ def save_checkpoint(
         # convert the state_dict back to hf format; do this inplace
         if self._model_type == ModelType.PHI3_MINI:
             state_dict[utils.MODEL_KEY] = phi3_tune_to_hf(state_dict[utils.MODEL_KEY])
+        elif self._model_type == ModelType.MISTRAL_REWARD:
```
Comment (kartikayk) [inline torchtune/utils/_checkpointing/_checkpointer.py:437]: same comment as above

File: torchtune/utils/_checkpointing/_checkpointer_utils.py (hunk 1 above)
Comment (kartikayk) [reply torchtune/utils/_checkpointing/_checkpointer_utils.py:27]: I think ```mistral_reward``` makes sense unless you think there will be classifier checkpoints we'll need to load?

Comment (SalmanMohammadi) [reply torchtune/utils/_checkpointing/_checkpointer_utils.py:27]: I'll leave as-is for now, we can always generalise later if there's use cases.

File: torchtune/utils/_checkpointing/_checkpointer.py (hunk 2 above)
Comment (SalmanMohammadi) [reply torchtune/utils/_checkpointing/_checkpointer.py:388]: I think this makes sense to me.

File: torchtune/models/convert_weights.py [hunk 4]
```diff
@@ -126,7 +126,7 @@ def hf_to_tune(
     repo in HF (https://huggingface.co/meta-llama/Llama-2-7b-hf).
 
     Args:
-        state_dict (Dict[str, torch.Tensor]): State dict in Meta's format.
+        state_dict (Dict[str, torch.Tensor]): State dict in HF's format.
```
Comment (kartikayk) [inline torchtune/models/convert_weights.py:129]: Nice catch!

**Review by kartikayk — APPROVED** [review]
Generally looks good, thank you!
